apis/python/src/tiledbsoma/fastercsx.cc

```diff
@@ -0,0 +1,621 @@
+
+    auto [n_row, n_col] = shape;
+    if (n_row < 0 || n_col < 0)
+        throw std::length_error("n_row and n_col must be >= 0");
+    auto n_major = (format == "csr") ? n_row : n_col;
+
+    if (major_idx_start < 0 || major_idx_end > n_major)
+        throw std::range_error(
+            "row_start must be >= 0 and row_end < array shape");
+    if (n_major != Bp.size() - 1)
+        throw std::length_error("n_rows does not match Bp.size()");
+    if (!out.writeable())
+        throw std::invalid_argument("out must be writable");
+    if (out.dtype().num() != Bd.dtype().num())
+        throw pybind11::type_error("out dtype must match Bd dtype");
+
+    // Get dispatch types
+    CsxIndexType csx_major_index_type = lookup_dtype_(
+        csx_index_type_dispatch, Bp.dtype(), "CSx indptr array");
+    CsxIndexType csx_minor_index_type = lookup_dtype_(
+        csx_index_type_dispatch, Bj.dtype(), "CSx indices array");
+    ValueType value_type = lookup_dtype_(
+        value_type_dispatch, Bd.dtype(), "CSx data array");
+
+    std::visit(
+        [&](auto value_type,
+            auto csx_major_index_type,
+            auto csx_minor_index_type) {
+            using CSX_MAJOR_INDEX =
+                typename decltype(csx_major_index_type)::type;
+            using CSX_MINOR_INDEX =
+                typename decltype(csx_minor_index_type)::type;
+            using VALUE = typename decltype(value_type)::type;
+
+            auto Bp_view = make_span_<CSX_MAJOR_INDEX>(Bp);
+            auto Bj_view = make_span_<CSX_MINOR_INDEX>(Bj);
+            auto Bd_view = make_casted_span_<VALUE, remap_value_t<VALUE>>(Bd);
+            auto out_view =
+                make_mutable_casted_span_<VALUE, remap_value_t<VALUE>>(out);
+
+            py::gil_scoped_release release;
+            return fastercsx::copy_to_dense(
+                ctx->thread_pool().get(),
+                major_idx_start,
+                major_idx_end,
+                shape,
+                cm_format,
+                Bp_view,
+                Bj_view,
+                Bd_view,
+                out_view);
+        },
+        value_type,
+        csx_major_index_type,
+        csx_minor_index_type);
+};
+
+#ifdef FASTERCSX__DEBUGGING_HOOKS
+void count_rows(
+    std::shared_ptr<tiledbsoma::SOMAContext> ctx,
+    const int64_t n_row,
+    const uint64_t nnz,
+    py::tuple Ai_,
+    py::array Bp) {
+    std::vector<py::array> Ai;
+    try  // convert py::tuple[py::array] to vector[py::array]
+    {
+        Ai = to_vector_<py::array>(Ai_);
+    } catch (const py::cast_error& e) {
+        throw pybind11::type_error(
+            std::string("All COO data must be tuple of ndarray (") + e.what() +
+            ")");
+    }
+
+    if (Bp.ndim() != 1)
+        throw std::range_error("All arrays must be of dimension rank 1.");
+
+    if (Bp.size() != (n_row + 1))
+        throw std::length_error("Pointer array size does not match n_rows.");
```

Comment:
Another `fmt::format` opportunity to give the caller more information about what they did wrong